mv: forbid IS NOT NULL on columns outside the primary key

cvybhu opened the pull request:
`IS NOT NULL` is currently allowed only when creating materialized views.
It's used to convey that the view will not include any rows that would make the view's primary key columns NULL.

Generally materialized views allow to place restrictions on the primary key columns, but restrictions on the regular columns are forbidden. The exception was IS NOT NULL - it was allowed to write regular_col IS NOT NULL. The problem is that this restriction isn't respected, it's just silently ignored (see #10365).

Supporting `IS NOT NULL` on regular columns seems to be as hard as supporting any other restrictions on regular columns.
It would be a big effort, and there is some reason why we don't support them.

For now let's forbid such restrictions, it's better to fail than be wrong silently.

Fixes: #10365

cvybhu commented:
The documentation will need to be updated, https://docs.scylladb.com/stable/cql/mv.html has the line:
> The columns that are not part of the base table primary key can only be restricted by an IS NOT NULL restriction. No other restriction is allowed.

nyh commented:
> The documentation will need to be updated, https://docs.scylladb.com/stable/cql/mv.html has the line:
> 
> > The columns that are not part of the base table primary key can only be restricted by an IS NOT NULL restriction. No other restriction is allowed.

The documentation is now part of the Scylla tree - the file you pointed to is docs/cql/mv.rst. So you can - and should - fix the documentation together with the code. Thanks.

cvybhu commented:
Will do.
Actually it looks like the documenation is also wrong at the moment, it says that only columns from the base table primary key can be restricted, but we allow restrictions on any columns from the view's primary key, not base table's.

cvybhu commented:
Hmm now that I think about it, maybe it would be better to put the check here: https://github.com/scylladb/scylladb/blob/20bff2bd1027b078987386d1fc8271b40ed9be07/cql3/statements/create_view_statement.cc#L278

I wouldn't need to add an argument to statement_restrictions and pass it around everywhere. I will try it out.

cvybhu commented:
v2:
* moved the check from `statement_restrictions` to `create_materialized_view`, it greatly simplified the change.
* updated the documentation to reflect the change

avikivity commented:
What happens if a cluster that has a view with an IS NOT NULL restriction is upgraded?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/4636/

nyh commented:
> > What happens if a cluster that has a view with an IS NOT NULL restriction is upgraded?
> 
> The check is done in `create_view_statement`, so if the view has already been created before the upgrade it would probably keep working. I'm not sure what would happen if a new node was added to the cluster. Is the schema recreated using multiple `CREATE x` statements?

I think the only place this is ever checked is in the CREATE VIEW call. If I remember correclty, we discovered all these weird issues when we noticed in Alternator the "IS NOT NULL" options were wrong in some cases, and nobody ever noticed because Alternator calls internal code to create a view, not the CQL "CREATE VIEW" code.

I honestly don't know what to do about the "backward compatibility" issue and don't know what to suggest. You saw this problem in the tests - people just added wrong IS NOT NULL in random places and the new code will break those old apps and their view-creation code. On the other hand, it's clear these applications didn't intend to do it in the first place :-(

nyh commented on cql3/restrictions/statement_restrictions.hh:
```diff
@@ -180,6 +180,10 @@ public:
         return _clustering_columns_restrictions;
     }
 
+    // Get a set of columns restrited by the IS NOT NULL restriction.
```
typo: restricted

nyh reviewed (approved):
Looks good to me, I like that we fix this (and the fix itself), but we need to be very aware that this make break broken applications which just used IS NOT NULL without thinking on everything.

One thing we can do is to "be bold" and make this change and tell users who complain that it's their fault (it really is). But @cvybhu, another thing we can do is to add a an option similar to how we added strict_allow_filtering - an option (say, call it "strict_mv_is_not_null") which today defaults to WARN - which warns (and logs?) if IS NOT NULL is used on the wrong thing - but in the future will be changed to TRUE to forbid IS NOT NULL on the wrong thing.

Anybody has an opinion what we should do here? @avikivity @tzach ? Adding an option is the "safest" and "most powerful" approach, but it's really ugly that we're collecting more of these options just to be backward-compatible with past bugs :-(

cvybhu commented:
Should I also modify `./conf/scylla.yaml`?

Looking at `strict_allow_filtering` it isn't present in `conf/scylla.yaml`, it's only present in `test/pylib/scylla_cluster.py::make_scylla_conf()`. Maybe that's a mistake, will the content generated by `make_scylla_conf()` end up in every node's `scylla.yaml`?

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1304/

nyh commented:
> Should I also modify `./conf/scylla.yaml`?
> 
> Looking at `strict_allow_filtering` it isn't present in `conf/scylla.yaml`, it's only present in `test/pylib/scylla_cluster.py::make_scylla_conf()`. Maybe that's a mistake, will the content generated by `make_scylla_conf()` end up in every node's `scylla.yaml`?

There's a philosophical question of whether every parameter which has a default value in the code should also appear in scylla.yaml, and if so, why do we even have a default value in the code?

A separate question is what is the point of adding these options if we never go back and change the default to enforcing instead of warning. We keep getting bug reports (e.g., https://github.com/scylladb/scylladb/issues/13533) because people get the wrong/bad behavior by default.

Obviously, test/pylib/scylla_cluster.py (and test/cql-pytest/run.py, which is similar and independent), aren't related to the defaults or scylla.yaml. In fact, I believe that the tests must run with the non-default enforcing mode, to be able to check it, because this enforcing mode is what we plan to be (eventually?) the default mode.

nyh commented on test/pylib/scylla_cluster.py:
```diff
@@ -96,6 +96,7 @@ def make_scylla_conf(workdir: pathlib.Path, host_addr: str, seed_addrs: List[str
         'request_timeout_in_ms': 300000,
 
         'strict_allow_filtering': True,
+        'strict_is_not_null_in_views': True,
```
Please also set this in test/cql-pytest/run.py.
It's an unfortunate duplication, but it exists.

cvybhu replied:
done

nyh commented on db/config.cc:
```diff
@@ -911,6 +911,7 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , cdc_dont_rewrite_streams(this, "cdc_dont_rewrite_streams", value_status::Used, false,
             "Disable rewriting streams from cdc_streams_descriptions to cdc_streams_descriptions_v2. Should not be necessary, but the procedure is expensive and prone to failures; this config option is left as a backdoor in case some user requires manual intervention.")
     , strict_allow_filtering(this, "strict_allow_filtering", liveness::LiveUpdate, value_status::Used, strict_allow_filtering_default(), "Match Cassandra in requiring ALLOW FILTERING on slow queries. Can be true, false, or warn. When false, Scylla accepts some slow queries even without ALLOW FILTERING that Cassandra rejects. Warn is same as false, but with warning.")
+    , strict_is_not_null_in_views(this, "strict_is_not_null_in_views", liveness::LiveUpdate, value_status::Used,db::tri_mode_restriction_t::mode::WARN, "In materialized views the IS NOT NULL restrictions works only on the view's primary key columns. On other columns the restriction isn't respected. Can be true, false, or warn. When false, Scylla accepts IS NOT NULL restrictions on regular columns ...
```
You mention scylla.yaml in the commit message, but I don't see it in this patch.
Your idea of putting "true" in the yaml and "warn" in the code is an interesting one, with a lot of appeal. The flip side is that a new cluster with an **existing** application will not work. Although perhaps this is good enough (when moving an existing application to a new cluster, presumably some period of instability and changes is to be expected).

avikivity replied:
Agree that a new cluster is better strict. Note that the failure will be very quick (as soon as the schema is created).

cvybhu replied:
> You mention scylla.yaml in the commit message, but I don't see it in this patch.

At first I based my changes on the ones for `strict_allow_filtering`, and there was no `strict_allow_filtering` in `scylla.yaml`, so I wasn't sure if I should add it, maybe it gets added automatically by some script (?).

I added my flag to `scylla.yaml` now. I assume that `strict_allow_filtering` should also be there, but we forgot to add it there. I will open an issue about it.

cvybhu replied:
Opened https://github.com/scylladb/scylladb/issues/14164

nyh commented on test/boost/restrictions_test.cc:
```diff
@@ -952,3 +952,99 @@ SEASTAR_THREAD_TEST_CASE(strict_allow_filtering_live_update) {
     }, cfg).get();
 }
+
+// Ok - only view primary key columns are restricted by IS NOT NULL
```
I wonder if these tests really belong in the "restrictions" file or the "materialized views" file, but I don't have a strong opinion, so I guess this is fine too.

cvybhu replied:
The tests for `strict_allow_filtering` are here, so I put my tests for `strict_is_not_null_in_views` below them.  ¯\_(ツ)_/¯

nyh reviewed (changes requested):
Looks good (thanks!), but I just have a couple of small suggestions

avikivity commented:
> Looks good to me, I like that we fix this (and the fix itself), but we need to be very aware that this make break broken applications which just used IS NOT NULL without thinking on everything.
> 
> One thing we can do is to "be bold" and make this change and tell users who complain that it's their fault (it really is). But @cvybhu, another thing we can do is to add a an option similar to how we added strict_allow_filtering - an option (say, call it "strict_mv_is_not_null") which today defaults to WARN - which warns (and logs?) if IS NOT NULL is used on the wrong thing - but in the future will be changed to TRUE to forbid IS NOT NULL on the wrong thing.
> 
> Anybody has an opinion what we should do here? @avikivity @tzach ? Adding an option is the "safest" and "most powerful" approach, but it's really ugly that we're collecting more of these options just to be backward-compatible with past bugs :-(

Agree with adding an option. "Be bold" works until the first user has an application crash.

avikivity commented on cql3/statements/alter_keyspace_statement.cc:
```diff
@@ -75,7 +75,7 @@ void cql3::statements::alter_keyspace_statement::validate(query_processor& qp, c
 
 future<std::pair<::shared_ptr<cql_transport::event::schema_change>, std::vector<mutation>>>
-cql3::statements::alter_keyspace_statement::prepare_schema_mutations(query_processor& qp, api::timestamp_type ts) const {
+cql3::statements::alter_keyspace_statement::prepare_schema_mutations(query_processor& qp, api::timestamp_type ts, std::vector<sstring>& warnings) const {
```
Is warnings an output parameter?

If so make it part of the output tuple.

cvybhu replied:
done

avikivity commented on cql3/statements/schema_altering_statement.cc:
```diff
@@ -70,12 +70,15 @@ schema_altering_statement::execute0(query_processor& qp, service::query_state& s
 
-            auto [ret, m] = co_await prepare_schema_mutations(qp, group0_guard.write_timestamp());
+            warnings.clear(); // clear the warnings before each attempt
+            auto [ret, m] = co_await prepare_schema_mutations(qp, group0_guard.write_timestamp(), warnings);
```
[ret, m, warning]

cvybhu replied:
done

avikivity commented:
How do upgrades work? If the bad IS NOT NULL is ignored early, then it won't be persisted in schema_tables.

But if it is, then we may have a problem.

cvybhu commented:
v2:
* Add `strict_is_not_null_in_views: true` in `scylla.yaml`, along with a description
* Make the flag description in `db/config.cc` nicer, it's based on the one in `scylla.yaml`.
* Return warnings from `prepare_schema_mutations` as the 3rd tuple element, instead of having a `std::vector<sstring>&` output argument. I also added an alias `cql3::cql_warnings_vec = std::vector<sstring>` to make it clear what the function is returning.
* Add the flag to `test/cql-pytest/run.py`

cvybhu commented:
> How do upgrades work? If the bad IS NOT NULL is ignored early, then it won't be persisted in schema_tables.
> 
> But if it is persisted, then we may have a problem.

I ran some tests manually:

### Test 1
* Start a single node with `strict_is_not_null_in_views: false`
* Create a keyspace, table, and a materialized view with an invalid `IS NOT NULL` restriction
* Stop the node
* Set `strict_is_not_null_in_views: true`
* Start the node
* Everything works, the view is fully usable, can be SELECTed, DROPped, etc.
* Trying to create a new materialized views with an invalid `IS NOT NULL` restriction throws an error, as expected

### Test 2
* Start a single node with `strict_is_not_null_in_views: false`
* Create a keyspace, table, and a materialized view with an invalid `IS NOT NULL` restriction
* Start a second node with `strict_is_not_null_in_views: true`, this node joins the first one
* Everything works, the view is fully usable, can be SELECTed, DROPped, etc.
* On the first node, creating a new invalid materialized view works, the other node accepts an invalid view from node1, even though it has `strict_is_not_null_in_views: true`
* On the second node, creating a new invalid materialized view fails with an error, as it has `strict_is_not_null_in_views: true`
* `ALTER KEYSPACE` works

I think it should work. The check happens only in `create_view_statement`, so as long as it's not used internally there shouldn't be any errors on upgrades, joins etc.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1684/

nyh commented on cql3/statements/create_view_statement.cc:
```diff
@@ -299,6 +301,35 @@ view_ptr create_view_statement::prepare_view(data_dictionary::database db) const
+        db.get_config().strict_is_not_null_in_views() == db::tri_mode_restriction_t::mode::WARN) {
+        sstring warning_text = fmt::format(
+            "The IS NOT NULL restriction is allowed only columns which are part of the view's primary key,"
+            " found columns: {}. Restrictions on these columns will be silently ignored. "
```
nitpick: "silently" isn't exactly the right word to use in a warning message ;-)

cvybhu replied:
Well, they "will be" ignored, but aren't at the moment :P

I could fix it, but I see that you already merged the PR, so I'll leave it as is.
